### PR TITLE
Mtom for binary encoded files

### PR DESCRIPTION
**Pull request by jonahkaye**
Ticket: #[1667](https://github.com/metriport/metriport-internal/issues/1667)

### Description

- adding mtom support for pdfs tiffs etc 
- refactoring code to use external libraries for mtom plumbing

### Testing

- Local
  - [x] lots of unit tests
  - [x] testing in pre prod env
- Staging
  - [ ] note that on staging implementers dont follow ihe standard and return non mtom response

### Release Plan

- [ ] Added to [monthly product update](https://www.notion.so/metriport/Customer-Updates-21b4e9d3ad5f4fd68db587a11db28cff?pvs=4) (introduce a feature that would be useful customers)
- [ ] Merge this

### CODE REVIEW

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts [hunk 1]
```diff
@@ -102,5 +105,128 @@ export async function sendSignedXmlMtom({
...
+    const binaryData: Buffer = Buffer.isBuffer(response.data)
+      ? response.data
+      : Buffer.from(response.data);
+    const mtomParts = await parseMTOMResp(binaryData, mtomContentType.boundary);
+    const soapData: Buffer = mtomParts.parts[0]?.body || Buffer.from("");
+    const fileData: Buffer = mtomParts.parts[1]?.body || Buffer.from("");
+    const fileType = detectFileType(fileData);
+    const fileType2 = detectFileType(soapData);
+    console.log("File type:", fileType);
+    console.log("File type:", fileType2);
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts:120]: the file types are right here but the fileData when written to a file or uploaded to s3 is corrupted

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts [hunk 2]
```diff
@@ -1,159 +1,90 @@
...
+            data = Buffer.from("");
+            break;
+          case "headerField":
+            headerName = buffer.slice(start, end).toString();
+            break;
+          case "headerValue":
+            headerValue = buffer.slice(start, end).toString();
+            break;
+          case "headerEnd":
+            //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:56]: can we get rid of this?

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts [hunk 3]
```diff
@@ -49,17 +49,58 @@ export function setS3UtilsInstance(s3Utils: S3Utils): void {
...
+
+function getMtomBytesAndMimeType(
+  documentResponse: DocumentResponse,
+  mtomResponse: IMTOMAttachments
+): { mimeType: string; decodedBytes: Buffer } {
+  if (documentResponseContainsMultipartCidReference(documentResponse)) {
+    const cid = getCidReference(documentResponse.Document.Include._href);
+    const attachment = getMtomAttachment(cid, mtomResponse);
+    const { mimeType, decodedBytes } = parseFileFromBuffer(attachment.body);
+    console.log("mimeType", mimeType);
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts:82]: leftover

File: packages/core/package.json [hunk 4]
```diff
@@ -84,6 +85,7 @@
     "playwright": "^1.39.0",
     "pvutils": "^1.1.3",
     "sequelize": "^6.37.1",
+    "whatwg-mimetype": "^4.0.0",
```
Comment (jonahkaye) [inline packages/core/package.json:88]: Web Hypertext Application Technology Working Group (WHATWG) library that implements https://mimesniff.spec.whatwg.org/

File: packages/core/package.json [hunk 5]
```diff
@@ -75,6 +75,7 @@
     "asn1js": "^3.0.5",
     "aws-sdk": "^2.1243.0",
     "dayjs": "^1.11.9",
+    "formidable": "^3.5.1",
```
Comment (jonahkaye) [inline packages/core/package.json:78]: https://www.npmjs.com/package/formidable

Comment (jonahkaye) [reply packages/core/package.json:78]: mtom parsing library

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts (hunk 2 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:56]: tried a bunch of things but the typescript compiler still freaks even if I have this

`if (resp.parts[partIndex]) {
              resp.parts[partIndex].body = data;
              partIndex++;
            }`

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/cid.ts [hunk 6]
```diff
@@ -1,9 +1,12 @@
...
   return cid.replace(cidPrefixRegex, "");
 }
 
-export function stripTags(content: string): string {
-  return content.replace(endTagRegex, "");
+function addTags(content: string): string {
+  return `<${content}>`;
+}
+
+export function getCidReference(cid: string): string {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/cid.ts:10]: https://datatracker.ietf.org/doc/html/rfc2392

`A "cid" URL is converted to the corresponding Content-ID message header by removing the "cid:" prefix, converting the % encoded character to their equivalent US-ASCII characters, and enclosing the remaining parts with an angle bracket pair, "<" and ">".`

Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/cid.ts:10]: this is from the RFC for content Ids in Multipart Messages

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:1]: mostly borrowed from https://github.com/vpulim/node-soap/blob/c40e5ea697cf9e4c9af637ed1f889e030fb8009a/src/utils.ts#L69
mit license. Not sure if we need an explicit declaration of anything here about that

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts [hunk 7]
```diff
@@ -114,6 +119,12 @@ export async function sendSignedXmlMtom({
       "Cache-Control": "no-cache",
     },
     httpsAgent: agent,
+    responseType: "arraybuffer",
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts:122]: very important. took 2 days to figure out but without this, axios automatically utf-8 encodes data which rsults in pdfs and tiffs getting corrupted

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/process-dr.test.ts [hunk 8]
```diff
@@ -67,129 +176,39 @@ describe("processDRResponse", () => {
...
         gateway: outboundDrRequest.gateway,
         outboundRequest: outboundDrRequest,
-        contentType: "application/xml",
       },
     });
     expect(response.operationOutcome?.issue[0]?.severity).toEqual("information");
   });
 });
-
-describe.skip("processDRResponse for various file types and verify successful upload without corruption", () => {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/process-dr.test.ts:111]: removing non local tests

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/process-mtom.test.ts
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/process-mtom.test.ts:1]: no longer a concept of mtom vs non-mtom tests. Everything now in process-dr-test.ts

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts (hunk 7 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts:122]: we never convert anything to strings now for DRs. Everything is buffers that never get encoded as strings ever!

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts [hunk 9]
```diff
@@ -77,14 +77,22 @@ export async function parseMtomResponse(
...
     throw new Error("Parsing of content type failed");
   }
   const boundary = parsedContentType.parameters.get("boundary");
-  if (!boundary) {
-    throw new Error("No boundary parameter found in content type.");
-  }
   return boundary;
 }
+
+export function convertSoapResponseToMtomResponse(buffer: Buffer): IMTOMAttachments {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:89]: so testing on staging, where implementers dont follow IHE spec, works. 
they return non mtom responses to mtom requests which is wrong but we should still handle it

**Review by RamilGaripov — COMMENTED** [review]
Left a few minor comments. 
This mostly looks good to me. But that part in `xca/mtom/parser` - I'd prefer that someone else checks that logic as well

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/process-dr.test.ts [hunk 10]
```diff
@@ -1,12 +1,123 @@
...
+        gateway: outboundDrRequestMtom.gateway,
+        outboundRequest: outboundDrRequestMtom,
+      },
+    });
+
+    expect(response).toBeDefined();
+    expect(response.documentReference).toHaveLength(3);
+    expect(response.documentReference?.[0]?.contentType).toBe(xmlFile.mimeType);
+    expect(response.documentReference?.[1]?.contentType).toBe(pdfFile.mimeType);
+    expect(response.documentReference?.[2]?.contentType).toBe(tiffFile.mimeType);
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/process-dr.test.ts:118]: nit: this requires files on a certain order; while it works w/ the current impl, if it its not a requirement/contract one could build the test making it so `response` contains doc refs w/ the expected contents; something like:

```ts
expect(response.documentReference).toBeTruthy();
expect(response.documentReference).toHaveLength(3);
const contentTypes = response.documentReference?.map(d => d.contentType);
expect(contentTypes).toEqual(
  expect.arrayContaining([xmlFile.mimeType, pdfFile.mimeType, tiffFile.mimeType])
);
expect(response.documentReference?.[0]?.contentType).toBe(xmlFile.mimeType);
```

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts
Comment (leite08) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:1]: MIT requires License and Copyright notice (https://github.com/vpulim/node-soap/blob/c40e5ea697cf9e4c9af637ed1f889e030fb8009a/LICENSE), so maybe have a dedicated file for what came from that repo with the notice header on that file?

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts [hunk 11]
```diff
@@ -49,17 +48,57 @@ export function setS3UtilsInstance(s3Utils: S3Utils): void {
...
+  return (
+    typeof documentResponse.Document !== "string" && !!documentResponse.Document?.Include?._href
+  );
+}
+
+function documentResponseContainsDocument(
+  documentResponse: DocumentResponse
+): documentResponse is DocumentResponse & { Document: string } {
+  return typeof documentResponse.Document === "string";
+}
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts:63]: nit: Both return true if the doc resp contains Document (and additional checks), maybe add the specifics on the name as well? Like:
- `documentResponseContainsMultipartCidReferenceDocument`
- `documentResponseContainsStringDocument`

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts [hunk 12]
```diff
@@ -150,27 +192,31 @@ async function handleSuccessResponse({
   }
 }
 
-export async function processDrResponseSoap({
-  drResponse: { response, success, gateway, outboundRequest },
+export async function processDrResponse({
+  drResponse: { errorResponse, mtomResponse, gateway, outboundRequest },
 }: {
   drResponse: DrSamlClientResponse;
 }): Promise<OutboundDocumentRetrievalResp> {
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts:199]: Why an object param if it only contains one property? Couldn't it just be

```ts
function processDrResponse({ 
  errorResponse, mtomResponse, gateway, outboundRequest
}: DrSamlClientResponse)...
```

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dr-requests.ts
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dr-requests.ts:1]: nit: on the capture.error we should decompose `extra`:

```ts
      capture.error(msg, {
        extra: {
          context,
          ...extra,
        }
```

**Review by leite08 — COMMENTED** [review]
A few comments, a few nits.

WRT to testing, I like that we're keeping tests up-to-date, just wondering about all those tests and respective XML files, are them all being addressed w/ the new tests?

File: packages/core/package.json (hunk 5 above)
Comment (jonahkaye) [reply packages/core/package.json:78]: afaik multer is primarily used  for uploading files via html forms as **multipart/form-data**. Mtom is **multipart/related** data. 
Multer doesn't look like it supports parsing multipart/related data (no mentions of multipart/related in the codebase or in issues or examples), but formidable does look like it supports what we are using multer for. 

But just fyi, multer is only being used for internal/admin logic in our codebase. I can make that ticket but not sure its worth the effort of replacing it.

File: packages/core/package.json (hunk 4 above)
Comment (jonahkaye) [reply packages/core/package.json:88]: The two libraries have different purposes. 
mime-types primarily provides functions to look up MIME types based on file extensions and vice versa.
whatwg-mimetype is a parser for arbitrary mime types. I looked at mime-types spec and didnt see anything that would be helpful for the specific way we are using it, namely to parse a contentType

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts (hunk 11 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts]: I changed to be more specific:

- `documentResponseContainsMultipartCidReferenceToDocument`
- `documentResponseContainsDocumentInSoapMessage`

**Review by RamilGaripov — APPROVED** [review]

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts (hunk 12 above)
Comment (leite08) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts:207]: This is when we have multiple parameters, not when there's only one. If we add more params, then we can update the function signature/contract.

**Comment by leite08** [comment]
> re: the two comments about using other npm packages, I investigated both and they don't meet our needs for this PR. We could swap out multer for formidable but imo it isnt worth ticketing since multer is only used in admin flows and we wont realistically ever resolve it since its not a high enough prio to ever get to the top of the stack

Can you share why didn't they meet our needs for this PR?

re: not worth creating a ticket

I see what you mean, but then if we just push stuff through with no regards to maintainability then we're not leaving the environment better than we found, right?

**Pull request merged**